### PR TITLE
Rewrite CreateNewBlock

### PR DESCRIPTION
**Pull request by morcos**
WARNING: This hasn't been used for mainnet mining yet as far as I know.

~~NOTE: This includes a commit which changes the default block priority size to 0.  This code has been optimized for not including transactions in a priority space.  If block priority space > 0 then the code does not offer too much of a performance improvement over the existing code.~~ (EDIT: now its fast regardless)

The mempool is explicitly assumed to be responsible for maintaining consistency of transactions with respect to not spending non-existent outputs, not double spending, script validity and coinbase maturity.  Only finality of transactions is checked before assembling a block.

A final call to TestBlockValidity is still performed.  If an invalid block has been constructed, then an error is thrown ~~logged and a NULL pointer is returned.~~ (EDIT: reverted to original behavior with more informative error)

This code produces the same blocks as the code it replaces with the following exceptions:
- The fee rate sort tie breaker is hash instead of priority
- The priority sort has a secondary tie breaker of hash
- ~~The priority block is not limited to transactons above AllowFree()~~ (EDIT: fixed)
- ~~The fee rate sorted part of the block is not limited to transactions above minRelayTxFee~~ (EDIT: fixed)
- (EDIT: The blocks are the same if you keep scanning to the end to try to fill up the last remaining space, but this code stops after trying 50 times to fill the last 1000 bytes)

Comparing this to the original code over 2000 calls to CreateNewBlock over the last 2 days.
(blockprioritysize=0, maxmempool=300, dbcache=1000, maxsigcachesize=1000000)
Time in ms

|  | Time to assemble block | Additional time to perform final TestBlockValidity |
| --- | --- | --- |
| master | 2602 | 240 |
| this pull | 14 | 438 |

The extra slowness in TestBlockValidity is because the cache used to be warmed up in the assembly code.

### CODE REVIEW

File: src/txmempool.h [hunk 1]
```diff
@@ -78,7 +80,8 @@ class CTxMemPoolEntry
 
 public:
     CTxMemPoolEntry(const CTransaction& _tx, const CAmount& _nFee,
-                    int64_t _nTime, double _dPriority, unsigned int _nHeight, bool poolHasNoInputsOf = false);
+                    int64_t _nTime, double _dPriority, unsigned int _nHeight, bool poolHasNoInputsOf = false,
+                    unsigned int nSigOps = 0);
```
Comment (luke-jr) [inline src/txmempool.h]: NACK defaulting nSigOps to 0 when we depend on it being accurate.

File: src/miner.cpp [hunk 2]
```diff
@@ -351,8 +296,10 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
         pblocktemplate->vTxSigOps[0] = GetLegacySigOpCount(pblock->vtx[0]);
 
         CValidationState state;
-        if (!TestBlockValidity(state, *pblock, pindexPrev, false, false))
-            throw std::runtime_error("CreateNewBlock(): TestBlockValidity failed");
+        if (!TestBlockValidity(state, *pblock, pindexPrev, false, false)) {
+            LogPrintf("ERROR: CNB TestBlockValidity failed: %s\n",FormatStateMessage(state));
+            return NULL;
```
Comment (luke-jr) [inline src/miner.cpp]: Why change exception to returning NULL here? IMO this should be an independent PR.

File: src/txmempool.h (hunk 1 above)
Comment (morcos) [inline src/txmempool.h]: ok i agree, it makes more sense to default it high.

Comment (luke-jr) [inline src/txmempool.h]: It does not make sense to default it period. Note the exact sigop count is part of the GBT response.

Comment (morcos) [inline src/txmempool.h]: OK agreed.  The default is just for the tests, I think for now it makes sense to remove all defaults from the constructor and just make the tests fill them in.  Unfortunately there are 2 other pulls open which also add arguments to the constructor and require changing the tests.  I'm going to hold off updating the tests in this pull until we're a bit closer assuming those might be merged first.

File: src/miner.cpp (hunk 2 above)
Comment (morcos) [inline src/miner.cpp]: The idea was this is now more likely to happen, and it seemed to meet that the JSON error from GetBlockTemplate would still be sufficient notification, and it would be better to not bring the node down.  However, I'm happy to leave the existing behavior if people think crashing the node is the right outcome.

**Comment by morcos** [comment]
OK, this has been rewritten to use `CTxMemPoolEntry::GetPriority` for the priority sort.
This will make it very fast even when maintaining a priority space in the block.  So the commit changing the default blockprioritysize has been removed.  

Of course until #6357 is merged or some other decision is taken regarding priority, it will not sort by an accurate measure of priority.

File: src/miner.cpp (hunk 2 above)
Comment (luke-jr) [inline src/miner.cpp]: Exceptions here should not crash the node... it just gets returned as a nicer JSON-RPC error.

**Comment by morcos** [comment]
I addressed @luke-jr 's comment about changing the method of failure return.  I still modified it to give some extra state info.   That commit and other fixes will get squashed.

Among the remaining outstanding items is to fix the constructor for CTxMemPoolEntry to not have any default arguments.  It would be nice to know a merge order since this will conflict with #6357 and #6915 and involves tedious updating of a bunch of tests.

**Comment by morcos** [comment]
I rebased this and squashed fixes for comments so far.

I rebased it on top of #7008, because I needed to pick some path forward on how priority would be addressed in this PR.  The first 3 commits are therefore part of that pull and not this one.  If we decide to merge #6357 as well, then it should be an easy change.

I will continue testing, but I don't have any other changes planned right now.

**Comment by morcos** [comment]
Thanks @sdaftuar for finding a bug in the way I was doing PrioritiseTransaction.   There should probably be RPC tests for those functions.

**Comment by morcos** [comment]
Apologies for all the churn.  Squashed again and rebased off a slightly modified index @sdaftuar created for #7062.   This introduces one more minor difference from the original code in that fee rates are compared as doubles now.  

If I correct for these differences:
- using a score index which compares fee rates not doubles
- make the original code use the hash tie breakers
- search up 5000 txs to fill the final 1000 bytes
- implement #6357 to correctly compute priority

Then I generate the exact same blocks as the old code modulo a few rare cases where double imprecision in the different priority calculation causes tx reordering in the priority space.

File: src/miner.cpp [hunk 3]
```diff
@@ -350,8 +287,9 @@ CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const CScript& s
         pblocktemplate->vTxSigOps[0] = GetLegacySigOpCount(pblock->vtx[0]);
 
         CValidationState state;
-        if (!TestBlockValidity(state, chainparams, *pblock, pindexPrev, false, false))
-            throw std::runtime_error("CreateNewBlock(): TestBlockValidity failed");
+        if (!TestBlockValidity(state, chainparams, *pblock, pindexPrev, false, false)) {
+            throw std::runtime_error("CreateNewBlock(): TestBlockValidity failed: " + FormatStateMessage(state));
```
Comment (jtimon) [inline src/miner.cpp]: can you ```strprintf("%: TestBlockValidity failed: %", __func__, ...``` instead?

**Comment by sdaftuar** [comment]
Addressed @jtimon's nit.

**Comment by btcdrak** [comment]
@morcos Needs rebase.

**Comment by jameshilliard** [comment]
TestBlockValidity should probably be done in a background thread so that it does not delay GBT since TestBlockValidity is just a sanity check.

**Comment by morcos** [comment]
rebased

**Comment by gmaxwell** [comment]
Concept ACK. Some tested-ack. Will test more now.

**Comment by jgarzik** [comment]
ut ACK

**Comment by laanwj** [comment]
Needs rebase again...

**Comment by morcos** [comment]
that was an expected merge conflict, will rebase shortly

@laanwj my warning about not having mined, was just a literal warning because i don't have sufficient hashpower to mine on mainnet.  I have created 1000s upon 1000s of block templates that all pass TestBlockValidity both in code running this pull live for the past 2 weeks (and prior iterations for weeks before that) and in historical simulation over 6 weeks of data.

**Comment by sipa** [comment]
ACK

**Comment by morcos** [comment]
rebased

**Comment by luke-jr** [comment]
This was not ready for merging. It needs #6357 first.

**Comment by laanwj** [comment]
> This was not ready for merging. It needs #6357 first.

#7008, which took the place of #6537 was merged
